### PR TITLE
Improve parsing performance

### PR DESCRIPTION
**Pull request by RoosterDragon**
- In LinguiniParser, ZeroCopyReader, ZeroCopyUtil: change helpers from dealing in ReadOnlyMemory<char> to char. In inner loops this removes a lot of redundant checks for a span of length 1 and inlines more readily.
- In ZeroCopyReader add helpers for SeekEol, IndexOfAnyChar. These use existing helper methods of Span that are SIMD accelerated to search for indexes quickly. These provide significant speedups compared to incrementing the index in a loop.
- In FluentBundle.AddResourceOverriding, provide a specialized 'override' method to avoid two dictionary lookups and allocating an empty list.

----

This makes parsing performance 1.5x faster. On my machine, the following test project runs in 2.9 seconds, compared to 4.5 seconds previously.

<details>

<summary>Test Project</summary>

```c#
using System;
using System.Diagnostics;
using System.Globalization;
using System.IO;
using System.Linq;
using Linguini.Bundle.Builder;
using Linguini.Syntax.Parser;

namespace ConsoleApp
{
    internal class Program
    {
        static void Main()
        {
            var sw = new Stopwatch();
            for (var i = 0; i < 10000; i++)
            {
                Setup(out var culture, out var streams);

                sw.Start();
                var bundle = LinguiniBuilder.Builder()
                    .CultureInfo(culture)
                    .SkipResources()
                    .SetUseIsolating(false)
                    .UseConcurrent()
                    .UncheckedBuild();

                foreach (var stream in streams)
                {
                    using var reader = new StreamReader(stream);
                    var parser = new LinguiniParser(reader);
                    var resource = parser.Parse();

                    bundle.AddResourceOverriding(resource);
                }
                sw.Stop();
            }

            sw.Stop();

            Console.WriteLine($"Done in {sw.Elapsed.TotalSeconds:0.00s}");
        }

        static void Setup(out CultureInfo culture, out MemoryStream[] streams)
        {
            culture = new CultureInfo("en");
            var paths = new[]
            {
                @"your-test-files-here.ftl",
            };
            streams = paths.Select(p => new MemoryStream(File.ReadAllBytes(p))).ToArray();
        }
    }
}
```

I used the following test files, but feel free to try with whatever you have to hand.
- https://github.com/OpenRA/OpenRA/blob/06df75ffee07c074f8ded8854cc9892db2d73cd2/mods/common/languages/en.ftl
- https://github.com/OpenRA/OpenRA/blob/06df75ffee07c074f8ded8854cc9892db2d73cd2/mods/common/languages/rules/en.ftl
- https://github.com/OpenRA/OpenRA/blob/06df75ffee07c074f8ded8854cc9892db2d73cd2/mods/ra/languages/rules/en.ftl
</details>

### CODE REVIEW

**Comment by Ygg01** [comment]
Seems to be failing tests. 

Also, some benchmarking would be nice. I can add some Benchmarks later (although benchmarking is a bit of a dark science - easy to get wrong).

**Comment by Ygg01** [comment]
Here are results for current master:

> BenchmarkDotNet=v0.13.5, OS=Windows 10 (10.0.19044.2965/21H2/November2021Update)
AMD Ryzen 9 5900X, 1 CPU, 24 logical and 12 physical cores
.NET SDK=7.0.102
  [Host] : .NET 5.0.12 (5.0.1221.52207), X64 RyuJIT AVX2 DEBUG

> Toolchain=InProcessEmitToolchain  


|          Method |  N |       Mean |     Error |    StdDev |
|---------------- |--- |-----------:|----------:|----------:|
| BenchmarkParser |  1 |   3.009 ms | 0.0236 ms | 0.0197 ms |
| BenchmarkParser | 25 |  74.498 ms | 0.3450 ms | 0.3227 ms |
| BenchmarkParser | 50 | 150.409 ms | 0.7432 ms | 0.6952 ms |

|          Method |  N |      Mean |     Error |    StdDev |
|---------------- |--- |----------:|----------:|----------:|
| BenchmarkParser |  1 |  1.337 ms | 0.0124 ms | 0.0110 ms |
| BenchmarkParser | 25 | 32.971 ms | 0.4621 ms | 0.4096 ms |
| BenchmarkParser | 50 | 65.290 ms | 0.7167 ms | 0.6354 ms |

All in all those are some nice speedups.

File: Linguini.Syntax/Parser/LinguiniParser.cs [hunk 1]
```diff
@@ -602,13 +598,22 @@ private bool TryGetTextSlice([NotNullWhen(true)] out TextSlice? textElement, out
             var startPos = _reader.Position;
             var textElementType = TextElementType.Blank;
 
-            while (_reader.TryPeekCharSpan(out var span))
+            int index;
+            while ((index = _reader.IndexOfAnyChar(" \n\r{}")) != -1)
```
Comment (Ygg01) [inline Linguini.Syntax/Parser/LinguiniParser.cs:602]: This needs to be a different function or to write is as such: 
```suggestion
            while ((index = _reader.IndexOfAnyChar(" \n\r{}".AsSpan())) != -1)
```

**Comment by Ygg01** [comment]
Ok, expect a release soonish. Need to write changelog.

**Comment by RoosterDragon** [comment]
@Ygg01 looks like the action failed partway through pushing to nuget, just a ping in case you hadn't noticed: https://github.com/Ygg01/Linguini/actions/runs/5237132620

**Comment by RoosterDragon** [comment]
May I request `Linguini.Bundle` as well?

**Pull request merged**